pallets/fusion/src/lib.rs

```diff
@@ -0,0 +1,1989 @@
+					amount,
+				});
+			}
+		});
+
+		Ok(())
+	}
+
+	/// Withdraw the fusion currency amount from the user's idle balance for a specific currency.
+	fn withdraw_from_currency_balance(
+		evm_address: EvmAddress,
+		currency_id: CurrencyId,
+		amount: FusionCurrencyBalance,
+	) -> DispatchResult {
+		FusionMemberCurrencyBalances::<T>::try_mutate(
+			evm_address,
+			currency_id,
+			|balance_opt| -> DispatchResult {
+				if let Some(balance) = balance_opt {
+					ensure!(
+						balance.amount >= amount,
+						Error::<T>::NotEnoughCurrencyBalanceForUser
+					);
+
+					balance.amount = balance.amount.saturating_sub(amount);
+
+					if balance.amount == 0 {
+						*balance_opt = None;
+					}
+				} else {
+					return Err(Error::<T>::NoCurrencyBalanceForUser.into());
+				}
+
+				Ok(())
+			},
+		)
+	}
+
+	/// Function to check if a pool should be removed and perform cleanup if necessary
+	fn check_and_cleanup_pool(pool: &FusionPool<T>) -> DispatchResult {
+		let pool_id = pool.pool_id;
+		let has_no_members = pool.members.is_empty();
+		let has_no_points = pool.total_staked_points == 0;
+		let has_no_staked_native = pool.total_staked_native == 0;
+		let has_no_unbonding_native = pool.total_unbonding_native == 0;
+		let has_no_rewards = !FusionEraRewards::<T>::iter_keys().any(|(_, id)| id == pool_id);
+		let has_no_exposures = !FusionExposures::<T>::iter_keys().any(|(_, id)| id == pool_id);
+
+		if has_no_members
+			&& has_no_points
+			&& has_no_staked_native
+			&& has_no_unbonding_native
+			&& has_no_rewards
+			&& has_no_exposures
+		{
+			FusionPools::<T>::remove(pool_id);
+			FusionPoolCount::<T>::mutate(|count| *count = count.saturating_sub(1));
+			Self::deposit_event(Event::PoolDeleted { pool_id });
+		}
+
+		Ok(())
+	}
+
+	/// Setup the fusion currency rates for the new era
+	fn setup_currency_rates(era: EraIndex) -> DispatchResult {
+		for (currency_id, currency) in FusionCurrencies::<T>::iter() {
+			// Skip if the currency is destroyed
+			if currency.is_destroyed {
+				continue;
+			}
+
+			// Try to get the new rate from the rate changes storage
+			let new_rate = FusionCurrencyRateChanges::<T>::get(currency_id).or_else(|| {
+				// Fallback to the current era's rate
+				FusionCurrencyRates::<T>::get(currency_id, era)
+			});
+
+			// If neither a new rate nor a current rate is found, trigger an error
+			let rate = new_rate.ok_or(Error::<T>::CurrencyRateNotFound)?;
+
+			// Insert the rate for the next era
+			FusionCurrencyRates::<T>::insert(currency_id, era + 1, rate);
+		}
+		Ok(())
+	}
+
+	/// Clean history depth storages and send old pending rewards to 'RewardRemainder'
+	fn clean_history_depth_storages(era: EraIndex) -> DispatchResult {
+		let history_depth = T::HistoryDepth::get();
+
+		if let Some(era_to_clear) = era.checked_sub(history_depth) {
+			// Clean fusion exposures
+			let pool_count = FusionPoolCount::<T>::get();
+			let _ = FusionExposures::<T>::clear_prefix(era_to_clear, pool_count, None);
```

Comment:
Here we don't need to use `FusionPoolCount` as the limit. We can just put `MAX` because at any given time we are not expecting to have more than 1000 pools or something like that.

Reply:
We have FusionPoolCount so why not use it ? You wanna save a read ?

Reply:
That would be great :)